Error: "invalid memory address or nil pointer dereference" when trying to add new node to existing cluster

Hi,

I'm currently experiencing an issue when trying to add a new node to an existing cluster. I'm running 2.0.3 from within a Docker container. My log looks somewhat like the following:


Feb 27 14:52:58 ip-10-21-1-94.ec2.internal systemd[1]: Starting etcd container...
Feb 27 14:52:58 ip-10-21-1-94.ec2.internal writer_process[4938]: etcd
Feb 27 14:52:58 ip-10-21-1-94.ec2.internal writer_process[4945]: etcd
Feb 27 14:52:58 ip-10-21-1-94.ec2.internal writer_process[4953]: Pulling repository quay.io/coreos/etcd
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4953]: Status: Image is up to date for quay.io/coreos/etcd:v2.0.3
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal systemd[1]: Started etcd container.
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for peers on http://0.0.0.0:2380
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for peers on http://0.0.0.0:7001
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for client requests on http://0.0.0.0:2379
Feb 27 14:52:59 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:52:59 etcd: listening for client requests on http://0.0.0.0:4001
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: could not get cluster response from http://10.21.1.43:2380: Get http://10.21.1.43:2380/members: net/http: request canceled while waiting for connection
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: name = ip-10-21-1-94.ec2.internal
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: data dir = /opt/etcd
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: member dir = /opt/etcd/member
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: heartbeat = 100ms
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: election = 1000ms
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: snapshot count = 10000
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: 2015/02/27 14:53:00 etcdserver: advertise client URLs = http://10.21.1.94:2379,http://10.21.1.94:4001
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: panic: runtime error: invalid memory address or nil pointer dereference
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: [signal 0xb code=0x1 addr=0x0 pc=0x4a5b5d]
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 16 [running]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.panic(0x7dc1e0, 0xa63493)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdserver.startNode(0xc20803a580, 0x0, 0x0, 0x0, 0xc208044f80, 0x0, 0x0, 0x10, 0x10)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:78 +0xdd
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdserver.NewServer(0xc20803a580, 0xc20803a580, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:171 +0x6bc
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdmain.startEtcd(0xc20807c000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:162 +0x1013
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/etcdmain.Main()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:58 +0x1c5
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: main.main()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/main.go:48 +0x1ab
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 19 [finalizer wait]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.park(0x4131d0, 0xa67fd8, 0xa66209)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.parkunlock(0xa67fd8, 0xa66209)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runfinq()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: runtime.goexit()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/proc.c:1445
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 20 [syscall]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: os/signal.loop()
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/os/signal/signal_unix.go:21 +0x1e
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by os/signal.init·1
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/os/signal/signal_unix.go:27 +0x32
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 28 [IO wait]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.runtime_pollWait(0x7fbec1e58938, 0x72, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/runtime/netpoll.goc:146 +0x66
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*pollDesc).Wait(0xc20807efb0, 0x72, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*pollDesc).WaitRead(0xc20807efb0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*netFD).Read(0xc20807ef50, 0xc208074000, 0x1000, 0x1000, 0x0, 0x7fbec1e573f0, 0xb)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/fd_unix.go:242 +0x34c
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net.(*conn).Read(0xc2080380a0, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/net.go:122 +0xe7
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/pkg/transport.timeoutConn.Read(0x7fbec1e58f30, 0xc2080380a0, 0x12a05f200, 0x12a05f200, 0xc208074000, 0x1000, 0x1000, 0x7fbec1e590b8, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /Users/unihorn/gopath/src/github.com/coreos/etcd/scripts/release/etcd/gopath/src/github.com/coreos/etcd/pkg/transport/timeout_conn.go:43 +0x16f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: github.com/coreos/etcd/pkg/transport.(*timeoutConn).Read(0xc2080405e0, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: <autogenerated>:21 +0xcc
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.noteEOFReader.Read(0x7fbec1e59068, 0xc2080405e0, 0xc20803a738, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:1203 +0x72
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*noteEOFReader).Read(0xc208040600, 0xc208074000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: <autogenerated>:124 +0xca
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: bufio.(*Reader).fill(0xc2080049c0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/bufio/bufio.go:97 +0x1b3
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: bufio.(*Reader).Peek(0xc2080049c0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/bufio/bufio.go:132 +0x101
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).readLoop(0xc20803a6e0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:782 +0x95
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:600 +0x93f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 26 [runnable]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).readLoop(0xc20803a2c0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:868 +0x829
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:600 +0x93f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 27 [select]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).writeLoop(0xc20803a2c0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:601 +0x957
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: goroutine 29 [select]:
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: net/http.(*persistConn).writeLoop(0xc20803a6e0)
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:885 +0x38f
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: created by net/http.(*Transport).dialConn
Feb 27 14:53:00 ip-10-21-1-94.ec2.internal writer_process[4961]: /usr/local/go/src/pkg/net/http/transport.go:601 +0x957
Feb 27 14:53:01 ip-10-21-1-94.ec2.internal systemd[1]: etcd.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Feb 27 14:53:01 ip-10-21-1-94.ec2.internal writer_process[5026]: etcd
Feb 27 14:53:01 ip-10-21-1-94.ec2.internal systemd[1]: Unit etcd.service entered failed state.


The existing cluster consists of 2 nodes, my etcd.service looks like this:


[Service]
EnvironmentFile=/etc/environment
TimeoutStartSec=0
SyslogIdentifier=writer_process
ExecStartPre=-/usr/bin/docker kill etcd
ExecStartPre=-/usr/bin/docker rm etcd
ExecStartPre=/usr/bin/docker pull quay.io/coreos/etcd:v2.0.3
ExecStart=/bin/bash -c "/usr/bin/docker run \
  -p 2379:2379 \
  -p 2380:2380 \
  -p 4001:4001 \
  -p 7001:7001 \
  --name etcd \
  -v /opt/etcd:/opt/etcd \
  -v /usr/share/ca-certificates/:/etc/ssl/certs \
  quay.io/coreos/etcd:v2.0.3 \
  -data-dir /opt/etcd \
  -name %H \
  -listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 \
  -advertise-client-urls http://10.21.1.94:2379,http://10.21.1.94:4001 \
  -listen-peer-urls http://0.0.0.0:2380,http://0.0.0.0:7001 \
  -initial-advertise-peer-urls http://10.21.1.94:2380,http://10.21.1.94:7001 \
  -initial-cluster-state existing \
  -initial-cluster ip-10-21-1-43.ec2.internal=http://10.21.1.43:2380,ip-10-21-1-43.ec2.internal=http://10.21.1.43:7001,ip-10-21-2-166.ec2.internal=http://10.21.2.166:2380,ip-10-21-2-166.ec2.internal=http://10.21.2.166:7001"
ExecStop=/usr/bin/docker kill etcd

I thought at first it might be related to the "could not get cluster response", but a quick check from .94 works:

core@ip-10-21-1-94 /etc/systemd/system $ curl -L http://10.21.1.43:2380/members
[{"id":12394471189341718283,"peerURLs":["http://10.21.2.166:2380","http://10.21.2.166:7001"],"name":"ip-10-21-2-166.ec2.internal","clientURLs":["http://10.21.2.166:2379","http://10.21.2.166:4001"]},{"id":13661735980988760828,"peerURLs":["http://10.21.1.43:2380","http://10.21.1.43:7001"],"name":"ip-10-21-1-43.ec2.internal","clientURLs":["http://10.21.1.43:2379","http://10.21.1.43:4001"]}]

Any help would be hugely appreciated.

Hi

I'm having the same problem when using -initial-cluster-state existing.

Double checked and no duplicates. 

2015/03/02 16:33:03 no data-dir provided, using default data-dir ./mojo5.etcd
2015/03/02 16:33:03 etcd: listening for peers on http://0.0.0.0:2380
2015/03/02 16:33:03 etcd: listening for client requests on http://0.0.0.0:2379
2015/03/02 16:33:03 etcd: listening for client requests on http://0.0.0.0:4001
2015/03/02 16:33:03 etcdserver: name = mojo5
2015/03/02 16:33:03 etcdserver: data dir = mojo5.etcd
2015/03/02 16:33:03 etcdserver: member dir = mojo5.etcd/member
2015/03/02 16:33:03 etcdserver: heartbeat = 100ms
2015/03/02 16:33:03 etcdserver: election = 1000ms
2015/03/02 16:33:03 etcdserver: snapshot count = 10000
2015/03/02 16:33:03 etcdserver: advertise client URLs = http://192.168.222.3:2379,http://192.168.222.3:4001
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4bf782]

goroutine 1 [running]:
github.com/coreos/etcd/etcdserver.startNode(0xc208046210, 0x0, 0x0, 0x0, 0xc208038820, 0x0, 0x0, 0x508760, 0xc208031750)
	/home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:78 +0xd2
github.com/coreos/etcd/etcdserver.NewServer(0xc208046210, 0xc2080703f0, 0x0, 0x0)
	/home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:171 +0x7cc
github.com/coreos/etcd/etcdmain.startEtcd(0xc208088000, 0x0, 0x0, 0x0)
	/home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:162 +0x15ee
github.com/coreos/etcd/etcdmain.Main()
	/home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:58 +0x1fc
main.main()
	/home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/main.go:48 +0x1de

goroutine 5 [syscall]:
os/signal.loop()
	/home/core/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/home/core/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 12 [runnable]:
net/http.(*persistConn).readLoop(0xc208046420)
	/home/core/go/src/net/http/transport.go:928 +0x9ce
created by net/http.(*Transport).dialConn
	/home/core/go/src/net/http/transport.go:660 +0xc9f

goroutine 13 [select]:
net/http.(*persistConn).writeLoop(0xc208046420)
	/home/core/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/home/core/go/src/net/http/transport.go:661 +0xcbc

goroutine 11 [sleep]:
net.func·019()
	/home/core/go/src/net/dnsclient_unix.go:240 +0x5a
created by net.loadConfig
	/home/core/go/src/net/dnsclient_unix.go:269 +0x20c

--- a/etcdserver/config.go
+++ b/etcdserver/config.go
@@ -46,9 +46,39 @@ type ServerConfig struct {
 	ElectionTicks int
 }
 
-// VerifyBootstrapConfig sanity-checks the initial config and returns an error
-// for things that should never happen.
-func (c *ServerConfig) VerifyBootstrapConfig() error {
+// VerifyBootstrapConfig sanity-checks the initial config for bootstrap case
+// and returns an error for things that should never happen.
+func (c *ServerConfig) VerifyBootstrap() error {
+	if err := c.verifyLocalMember(); err != nil {
+		return err
+	}
+	if err := c.Cluster.Validate(); err != nil {
+		return err
+	}
+	if c.Cluster.String() == "" && c.DiscoveryURL == "" {
+		return fmt.Errorf("initial cluster unset and no discovery URL found")
+	}
+	return nil
+}
+
+// VerifyJoinExisting sanity-checks the initial config for join existing cluster
+// case and returns an error for things that should never happen.
+func (c *ServerConfig) VerifyJoinExisting() error {
+	if err := c.verifyLocalMember(); err != nil {
+		return err
+	}
+	if err := c.Cluster.Validate(); err != nil {
+		return err
+	}
+	if c.DiscoveryURL != "" {
+		return fmt.Errorf("discovery URL should not be set when joining existing initial cluster")
+	}
+	return nil
+}
+
+// verifyLocalMember verifies that the local member is valid and is listed
+// in the cluster correctly.
+func (c *ServerConfig) verifyLocalMember() error {
 	m := c.Cluster.MemberByName(c.Name)
 	// Make sure the cluster at least contains the local server.
 	if m == nil {
@@ -56,14 +86,6 @@ func (c *ServerConfig) VerifyBootstrapConfig() error {
 	}
 	if uint64(m.ID) == raft.None {
 		return fmt.Errorf("cannot use %x as member id", raft.None)
-	}
-
-	if c.DiscoveryURL == "" && !c.NewCluster {
-		return fmt.Errorf("initial cluster state unset and no wal or discovery URL found")
-	}
-
-	if err := c.Cluster.Validate(); err != nil {
-		return err
 	}
 
 	// Advertised peer URLs must match those in the cluster peer list

--- a/etcdserver/config_test.go
+++ b/etcdserver/config_test.go
@@ -29,74 +29,76 @@ func mustNewURLs(t *testing.T, urls []string) []url.URL {
 	return u
 }
 
-func TestBootstrapConfigVerify(t *testing.T) {
+func TestConfigVerifyBootstrapWithoutClusterAndDiscoveryURLFail(t *testing.T) {
+	cluster, err := NewClusterFromString("", "")
+	if err != nil {
+		t.Fatalf("NewClusterFromString error: %v", err)
+	}
+	c := &ServerConfig{
+		Name:         "node1",
+		DiscoveryURL: "",
+		Cluster:      cluster,
+	}
+	if err := c.VerifyBootstrap(); err == nil {
+		t.Errorf("err = nil, want not nil")
+	}
+}
+
+func TestConfigVerifyExistingWithDiscoveryURLFail(t *testing.T) {
+	cluster, err := NewClusterFromString("", "node1=http://127.0.0.1:2380")
+	if err != nil {
+		t.Fatalf("NewClusterFromString error: %v", err)
+	}
+	c := &ServerConfig{
+		Name:         "node1",
+		DiscoveryURL: "http://127.0.0.1:4001/abcdefg",
+		PeerURLs:     mustNewURLs(t, []string{"http://127.0.0.1:2380"}),
+		Cluster:      cluster,
+		NewCluster:   false,
+	}
+	if err := c.VerifyJoinExisting(); err == nil {
+		t.Errorf("err = nil, want not nil")
+	}
+}
+
+func TestConfigVerifyLocalMember(t *testing.T) {
 	tests := []struct {
 		clusterSetting string
-		newclst        bool
 		apurls         []string
-		disc           string
 		shouldError    bool
 	}{
 		{
 			// Node must exist in cluster
 			"",
-			true,
 			nil,
-			"",
 
 			true,
 		},
 		{
-			// Cannot have duplicate URLs in cluster config
-			"node1=http://localhost:7001,node2=http://localhost:7001,node2=http://localhost:7002",
-			true,
-			nil,
-			"",
-
-			true,
-		},
-		{
-			// Node defined, ClusterState OK
+			// Initial cluster set
 			"node1=http://localhost:7001,node2=http://localhost:7002",
-			true,
 			[]string{"http://localhost:7001"},
-			"",
 
 			false,
 		},
 		{
-			// Node defined, discovery OK
-			"node1=http://localhost:7001",
-			false,
-			[]string{"http://localhost:7001"},
-			"http://discovery",
+			// Default initial cluster
+			"node1=http://localhost:2380,node1=http://localhost:7001",
+			[]string{"http://localhost:2380", "http://localhost:7001"},
 
 			false,
-		},
-		{
-			// Cannot have ClusterState!=new && !discovery
-			"node1=http://localhost:7001",
-			false,
-			nil,
-			"",
-
-			true,
 		},
 		{
 			// Advertised peer URLs must match those in cluster-state
 			"node1=http://localhost:7001",
-			true,
 			[]string{"http://localhost:12345"},
-			"",
 
 			true,
 		},
 		{
 			// Advertised peer URLs must match those in cluster-state
 			"node1=http://localhost:7001,node1=http://localhost:12345",
-			true,
 			[]string{"http://localhost:12345"},
-			"",
 
 			true,
 		},
@@ -108,15 +110,13 @@ func TestBootstrapConfigVerify(t *testing.T) {
 			t.Fatalf("#%d: Got unexpected error: %v", i, err)
 		}
 		cfg := ServerConfig{
-			Name:         "node1",
-			DiscoveryURL: tt.disc,
-			Cluster:      cluster,
-			NewCluster:   tt.newclst,
+			Name:    "node1",
+			Cluster: cluster,
 		}
 		if tt.apurls != nil {
 			cfg.PeerURLs = mustNewURLs(t, tt.apurls)
 		}
-		err = cfg.VerifyBootstrapConfig()
+		err = cfg.verifyLocalMember()
 		if (err == nil) && tt.shouldError {
 			t.Errorf("%#v", *cluster)
 			t.Errorf("#%d: Got no error where one was expected", i)

--- a/etcdserver/server.go
+++ b/etcdserver/server.go
@@ -155,6 +155,9 @@ func NewServer(cfg *ServerConfig) (*EtcdServer, error) {
 
 	switch {
 	case !haveWAL && !cfg.NewCluster:
+		if err := cfg.VerifyJoinExisting(); err != nil {
+			return nil, err
+		}
 		existingCluster, err := GetClusterFromRemotePeers(getRemotePeerURLs(cfg.Cluster, cfg.Name), cfg.Transport)
 		if err != nil {
 			return nil, fmt.Errorf("cannot fetch cluster info from peer urls: %v", err)
@@ -168,7 +171,7 @@ func NewServer(cfg *ServerConfig) (*EtcdServer, error) {
 		cfg.Print()
 		id, n, s, w = startNode(cfg, nil)
 	case !haveWAL && cfg.NewCluster:
-		if err := cfg.VerifyBootstrapConfig(); err != nil {
+		if err := cfg.VerifyBootstrap(); err != nil {
 			return nil, err
 		}
 		m := cfg.Cluster.MemberByName(cfg.Name)

